File: apps/messenger/src/modules/discord/models/discord-message-author.dto.ts
import { DiscordMessageAuthorDTOOptions } from "../types/discord-message-author-dto-options.type";

export class DiscordMessageAuthorDTO {
  public readonly avatarPngUrl?: string;
  public readonly bannerPngUrl?: string;
  public readonly bot: boolean;
  public readonly createdAt: Date;
  public readonly discriminator: string;
  public readonly displayName: string;
  public readonly globalName?: string;
  public readonly id: string;
  public readonly tag: string;
  public readonly system: boolean;
  public readonly username: string;

  public constructor(options: DiscordMessageAuthorDTOOptions) {
    this.avatarPngUrl = options.avatarPngUrl;
    this.bannerPngUrl = options.bannerPngUrl;
    this.bot = options.bot;
    this.createdAt = options.createdAt;
    this.discriminator = options.discriminator;
    this.displayName = options.displayName;
    this.globalName = options.globalName;
    this.id = options.id;
    this.tag = options.tag;
    this.system = options.system;
    this.username = options.username;
  }
}

File: apps/builder/src/main/kotlin/dev/felipeflohr/w2gservices/builder/dto/DiscordMessageAuthorDTO.kt
package dev.felipeflohr.w2gservices.builder.dto

import dev.felipeflohr.w2gservices.builder.annotations.NoArg
import dev.felipeflohr.w2gservices.builder.entities.DiscordMessageAuthorEntity
import java.util.Date

@NoArg
data class DiscordMessageAuthorDTO(
    var avatarPngUrl: String?,
    var bannerPngUrl: String?,
    var bot: Boolean,
    var createdAt: Date,
    var discriminator: String,
    var displayName: String,
    var globalName: String?,
    var id: String,
    var tag: String,
    var system: Boolean,
    var username: String,
) {
    fun toEntity(): DiscordMessageAuthorEntity {
        return DiscordMessageAuthorEntity(
            system = system,
            authorId = id,
            bot = bot,
            username = username,
            autCreatedAt = createdAt,
            avatarPngUrl = avatarPngUrl,
            bannerPngUrl = bannerPngUrl,
            globalName = globalName,
            displayName = displayName,
            id = null,
            message = null,
        )
    }

    fun toEntity(id: Long): DiscordMessageAuthorEntity {
        val entity = toEntity()
        entity.id = id
        return entity
    }
}
